R4R ADR 007 / 008: Specialization Groups Abstraction / CERT Group (#4827)
cosmos/cosmos-sdk: docs/architecture/adr-008-dCERT-group.md

# ADR 008: Decentralized Computer Emergency Response Team (dCERT) Group

## Changelog

- 2019 Jul 31 : Initial Conception

## Context

In order to reduce the number of parties involved with handling sensitive
information in an emergency scenario, we propose the creation of a
specialization group named The Decentralized Computer Emergency Response Team
(dCERT).  Initially this group's role is intended to serve as coordinators
between the validators, bug-hunters, developers, and other members of the wider
community.  During a crisis time, the dCERT group would facilitate and relay
input from a variety of stakeholders to the developers who are actively
devising a patch to the software, this way sensitive information does not need

> to be publicly disclosed while as some input from the community can still be

Reviewer: ```suggestion
to be publicly disclosed while some input from the community can still be
```

as defined in [ADR 007](./adr-007-specialization-groups.md). This will include the 
implementation of: 
 - continuous voting
 - slashing due to breach of soft contract
 - revoking a member due to breach of soft contract
 - compensation stipend from the community pool or other means decided by
   governance

This system necessitates the following new parameters: 
 - blockly stipend allowance per dCERT member (suggested ~$10K/yr of atoms)
 - maximum number of dCERT members (suggested 7) 
 - required staked slashable tokens for each dCERT member (suggested $10K of atoms)
 - unbonding time for dCERT staked token (suggested 3 week)
 - quorum for suspending a particular member (suggested 5/7) 

These parameters are expected to be implemented through the param keeper such 
that governance may change them at any given point. 

### Continuous Voting Electionator

An `Electionator` object is to be implemented with continuous voting is to be
implemented with the following specifications:
 - All delegation addresses may submit votes at any point which updates their 
   preferred representation on the dCERT group. 

>  - Preferred representation may be arbitrarily split between address (ex. 50%

Reviewer: Should be able to determine different voting power for different types of decisions. For example, Member A is security expert and gets outsize share of voting power on circuit-breaking, security-patch upgrades but has equal voting power on non-security issues

Author: As mentioned in 007:

> This type of community organization paves the way for
individual stakeholders to delegate votes by issue type, if in the future
governance proposals include a field for issue type.

I think for the time being this addition would be an overly complexity. We've made simplifications like this all throughout the hub for instance currently delegators don't even get to choose different validators between who secures their stake vs who's vote they automatically vote with when they don't vote for a proposal themselves (although arguably it would be beneficial to create this separation). 

In the future, I could imagine further specification to special permissions within the CERT group as a good idea, or actually even better distinct CERT groups with distinct permissions which need to coordinate with each other, but I think for the time being, equal voting power for everything involved in the CERT group makes the most sense to minimize complexity (and maximize comprehensibility to the community as we're rolling this feature out)

   to John, 25% to Sally, 25% to Carol) 

>  - Addresses which control the greatest amount of preferred-representation are

Reviewer: Will the lowest preferred member that is already in dCERT group get kicked if we reach maximum number?

Author: correct - I will note this in the documentation

 - vote to suspend a dCERT member. 

Here circuit-breaking refers to the capability to disable a groups of messages,
This could for instance mean: "disable all staking-delegation messages", or
"disable all distribution messages". This could be accomplished by verifying
that the message route has not been "circuit-broken" at CheckTx time (in
`baseapp/baseapp.go`). 

"unbreaking" a circuit is anticipated only to occur during a hard fork upgrade
meaning that no capability to unbreak a message route on a live chain is
required. 

Note also, that if there was a problem with governance voting (for instance a
capability to vote many times) then governance would be broken and should be
halted with this mechanism, it would be then up to the validator set to
coordinate and hard-fork upgrade to a patched version of the software where
governance is re-enabled (and fixed). If the dCERT group abuses this privilege
they should all be severely slashed.

## Status

> Proposed

## Consequences

### Positive

 - Potential to reduces the number of parties to coordinate with during an emergency 
   - Reduction in possibility of disclosing sensitive information to malicious parties

### Negative

>  - Centralization risks

Reviewer: I would like to see more specific issues here with centralization and how the current scheme will mitigate these concerns

Author: Have any specific ideas yourself btw?

Reviewer: I think the above section on governance is the main threat. Im still a bit fuzzy on how the attack would be mitigated tho. Here's a concrete example:

1. dCERT group disables governance at block N
2. dCERT group uses its privileges to spend all of the CommunityFund pool to themselves at block N+1.
3. What happens next?? If the chain hard-forks (starting from what block?) and slashes dCERT group, does dCERT group still have the funds that they stole from CommunityFund? Will the slashing occur through governance on the new chain or will the slashing penalty be encoded into the new genesis?

Author: Number 2 doesn't make _any_ sense. the dCERT group is payed a small pre-defined stipend, they most certainly should have control over the funds. IMO I don't think there should be _any_ group who holds privileges over choosing how to spend the community pool besides the entire governance. Where did you get the idea that the dCERT group would have that privilege? is something incorrectly worded to convey that idea?

Reviewer: Number 2 was just an example i made up of something a dCERT group might have control over and use maliciously.

Perhaps I'm not clear about what I would like clarified.

Let's say a malicious dCERT group successfully circuit breaks governance and then uses their privileges for personal gain since governance cannot overrule them.

The solution to this is that the validators hard-fork and slash the dCERT group. 

Will the validators hard-fork from before the dCERT group exploited their privileges? My understanding is that this would cause validators to double-sign since they already signed for the height where dCERT group was malicious.

Maybe the example step-by-step I made above doesn't make sense. But I would like to see some example process flow of what would happen in the case of dCERT group malicious activity to clarify things. Something like:

1. dCERT group circuitbreaks governance at height N
2. They do bad stuff at height N+1
3. Validators hard-fork (at what height?)
4. dCERT group gets slashed on new chain

This would help understanding the cost-benefit for a dCERT group being malicious.

If there's a way for validators to hard-fork from before any malicious activity gets committed, then there's virtually no benefit to being malicious.
If validators have to hard-fork for latest height on old chain, then the cost benefit is `ProfitFromAttack - SlashingPenalty`.

Let me know if my thinking on this is wrong

Author: Pretty sure that validators can just hard fork from before malicious activity, there shouldn't be double signing issues, because even though there are duplicate heights, there are on two different chains, hence, they are signing different blocks. This is relevant for the first hard-fork which reset the height to 0 however there was no possibility of double signing. 

Not to mention, in a real emergency situation the community _could_ make arbitrary changes to the distribution, such as retroactively rolling back transactions from dirty malicious money. 

I think this answers your question (does it?)

Reviewer: Yes it does. Though this does mess with the finality promise of a tendermint chain, it should happen so rarely i don't think it's a problem

Reviewer 2: I'd like to un-resolve this convo. I still think there is more discussion to be had here. We should really think about the possibly exploits and failure modes In a situation where the group decides to shut down governance messages. The coordination would be a huge nightmare and headache amongst major stakeholders (e.g. exchanges) to quickly hard-fork out this malicious group.

# ADR 008: Decentralized Computer Emergency Response Team (dCERT) Group

## Changelog

> - 2019 Jul 31 : Initial Conception

Reviewer 2: ```suggestion
- 2019 Jul 31: Initial Draft
```

## Context

> In order to reduce the number of parties involved with handling sensitive

Reviewer 3: This paragraph seems to be discussing responsibilities which are off-chain. Does it need to be in this ADR?

My two cents - ADRs should focus on technical implementations of particular ideas, not on the general merits of the ideas, especially when the ideas concern off-chain social structures.

information in an emergency scenario, we propose the creation of a
specialization group named The Decentralized Computer Emergency Response Team
(dCERT).  Initially this group's role is intended to serve as coordinators
between the validators, bug-hunters, developers, and other members of the wider

> community.  During a crisis time, the dCERT group would facilitate and relay

Reviewer 2: ```suggestion
community.  During a time of crisis, the dCERT group would aggregate and relay
```

input from a variety of stakeholders to the developers who are actively
devising a patch to the software, this way sensitive information does not need
to be publicly disclosed while some input from the community can still be
gained. 

> Additionally, a special privilege is proposed for the dCERT group:

Reviewer 3: This is the main purpose of the ADR, presumably?

the capacity to "circuit-break" (aka. temporarily disable)  a particular
message path. 

In the future it is foreseeable that the community may wish to expand the roles
of dCERT with further responsibilities such as the capacity to "pre-approve" a

> security update hard-fork on behalf of the community prior to a full community

Reviewer 4: ```suggestion
security update on behalf of the community prior to a full community
```

 - continuous voting
 - slashing due to breach of soft contract
 - revoking a member due to breach of soft contract
 - emergency disband of the entire dCERT group (ex. for colluding maliciously) 
 - compensation stipend from the community pool or other means decided by
   governance

This system necessitates the following new parameters: 
 - blockly stipend allowance per dCERT member (suggested ~$10K/yr of atoms)
 - maximum number of dCERT members (suggested 7) 
 - required staked slashable tokens for each dCERT member (suggested $10K of atoms)
 - unbonding time for dCERT staked token (suggested 3 week)
 - quorum for suspending a particular member (suggested 5/7) 
 - proposal wager for disbanding the dCERT group (suggested ~$100000 USD worth of atoms)

These parameters are expected to be implemented through the param keeper such 
that governance may change them at any given point. 

### Continuous Voting Electionator

An `Electionator` object is to be implemented with continuous voting is to be
implemented with the following specifications:
 - All delegation addresses may submit votes at any point which updates their 
   preferred representation on the dCERT group. 

>  - Preferred representation may be arbitrarily split between address (ex. 50%

Reviewer 2: ```suggestion
 - Preferred representation may be arbitrarily split between addresses (ex. 50%
```

   to John, 25% to Sally, 25% to Carol) 

>  - Addresses which control the greatest amount of preferred-representation are

Reviewer 2: How do you handle splits?

Reviewer: What happens if the community is split between two candidates at the lowest position and it turns contentious.

Both sides continue adding votes to one side of the election or the another, making the last position in the dCERT group see-saw between both candidates frequently (maybe even every couple of blocks). If there's a continuous election, i could see this potentially happening

Author: @AdityaSripal good point we could add a stabilizing time period, such that you can only be replaced if you've been on dCERT for one month. with would, in the worst case produce a 1 month see-saw

Reviewer 2: We should amend to include this

Author: yeah I'm going to add into the ADR with an enable parameter, such that this stabilizer could be enabled be governance post-mortem

Author: Added this discussion!

   eligible to join the dCERT group (up the _maximum number of dCERT members_)
 - In order for a new member to be added to the dCERT group they must 
   send a transaction accepting their admission at which point the validity of
   their admission is to be confirmed. 

>  - If the dCERT group is already full and new member is admitted, the existing

Reviewer 2: ditto

   dCERT member with the lowest amount of atoms is kicked from the dCERT group. 

### Staking/Slashing

All members of the dCERT group must stake tokens _specifically_ to maintain
eligibility as a dCERT member. This staking mechanism must be designed with an
unbonding period. Slash a particular dCERT member due to soft-contract breach
should be performed by governance on a per member basis based on the magnitude

> of the breach.  The process flow is anticipated to be that a dCERT member is

Reviewer 2: How are they suspended? What if they breach contract but a governance proposal claiming they've done so isn't submitted until right before their unbonding period is over?

suspended by the dCERT group prior to being slashed by governance. 

Additionally in the case of an emergency situation of a colluding and malicious
dCERT group, the community needs the capability to disband the entire dCERT
group and likely fully slash them. This could be achieved though a special new

> proposal type (general governance) which would halt the functionality of the

Reviewer 2: not sure "general governance" is needed here. I imagine this module will simply implement such a proposal type.

> dCERT group until the proposal was concluded. This special proposal type would

Reviewer 2: Why? If the proposal creator was malicious or simply wanted to see the entire group disbanded, then it would be highly unlikely for a quorum of stakeholders to vote on this. I don't see the need for anything special outside of a custom governance proposal type.

likely need to also have a fairly large wager which could be slashed if the
proposal creator was malicious. 

### dCERT membership transactions

Active dCERT members 
 - change of the description of the dCERT group
 - circuit break a message route
 - vote to suspend a dCERT member. 

Here circuit-breaking refers to the capability to disable a groups of messages,

> This could for instance mean: "disable all staking-delegation messages", or

Reviewer 2: idk...this seems risky to have 7 members have the ability at will to just circuit break any message. If they do so, then the stakeholders would have to submit a disbandment governance proposal which takes time. Am I missing something here?

"disable all distribution messages". This could be accomplished by verifying
that the message route has not been "circuit-broken" at CheckTx time (in
`baseapp/baseapp.go`). 

> "unbreaking" a circuit is anticipated only to occur during a hard fork upgrade

Reviewer 2: What if the group (or a cabal in the group) does so maliciously?

meaning that no capability to unbreak a message route on a live chain is
required. 

Note also, that if there was a problem with governance voting (for instance a
capability to vote many times) then governance would be broken and should be
halted with this mechanism, it would be then up to the validator set to
coordinate and hard-fork upgrade to a patched version of the software where
governance is re-enabled (and fixed). If the dCERT group abuses this privilege
they should all be severely slashed.

## Status

> Proposed

## Consequences

### Positive

 - Potential to reduces the number of parties to coordinate with during an emergency 

>    - Reduction in possibility of disclosing sensitive information to malicious parties

Reviewer 2: ```suggestion
- Reduction in possibility of disclosing sensitive information to malicious parties
```

# ADR 008: Decentralized Computer Emergency Response Team (dCERT) Group

## Changelog

- 2019 Jul 31: Initial Draft

## Context

In order to reduce the number of parties involved with handling sensitive
information in an emergency scenario, we propose the creation of a
specialization group named The Decentralized Computer Emergency Response Team
(dCERT).  Initially this group's role is intended to serve as coordinators
between various actors within a blockchain community such as validators,

> bug-hunters, developers.  During a time of crisis, the dCERT group would

Reviewer 2: ```suggestion
bug-hunters, and developers.  During a time of crisis, the dCERT group would
```

aggregate and relay input from a variety of stakeholders to the developers who
are actively devising a patch to the software, this way sensitive information
does not need to be publicly disclosed while some input from the community can
still be gained. 

Additionally, a special privilege is proposed for the dCERT group:

> the capacity to "circuit-break" (aka. temporarily disable)  a particular

Reviewer 2: I believe we talked about initially having the implementation _not_ include this functionality but which can later be easily added. I think we should make note of that here 👍

Author: Added the text

> Note that this privilege should be enabled/disabled globally with a governance
parameter such that this privilege could start disabled and later be enabled
through a parameter change proposal, once a dCERT group has been established.

message path. 

In the future it is foreseeable that the community may wish to expand the roles
of dCERT with further responsibilities such as the capacity to "pre-approve" a
security update on behalf of the community prior to a full community
wide vote whereby the sensitive information would be revealed prior to a
vulnerability being patched on the live network.  

## Decision

The dCERT group is proposed to include an implementation of a `SpecializationGroup`
as defined in [ADR 007](./adr-007-specialization-groups.md). This will include the 
implementation of: 
 - continuous voting
 - slashing due to breach of soft contract
 - revoking a member due to breach of soft contract
 - emergency disband of the entire dCERT group (ex. for colluding maliciously) 
 - compensation stipend from the community pool or other means decided by
   governance

This system necessitates the following new parameters: 
 - blockly stipend allowance per dCERT member 
 - maximum number of dCERT members 

>  - required staked slashable tokens for each dCERT member 

Reviewer 2: What happens when this dips below threshold?

Author: a dCERT member can _be_ a member if it has the required tokens staked _in_ the dCERT algorithm. If those tokens are unbonded then the dCERT member is kicked from the group.

^ added this text under `Staking/Slashing` section in this ADR

>  - unbonding time for dCERT staked token 

Reviewer 2: How is this different than the global unbonding time?

Author: It should probably be the same, I've modified text throughout to reflect this and deleted this bullet point. I guess though during implementation we will probably need another parameter because we do not yet have cross module parameters :( 

CC https://github.com/cosmos/cosmos-sdk/issues/3472

 - quorum for suspending a particular member 
 - proposal wager for disbanding the dCERT group 

These parameters are expected to be implemented through the param keeper such 
that governance may change them at any given point. 

### Continuous Voting Electionator

> An `Electionator` object is to be implemented with continuous voting is to be

Reviewer 2: This sentence reads funny

Author: wow you're right. Changed to: 

```
An `Electionator` object is to be implemented as continuous voting and with the
following specifications:
```

implemented with the following specifications:
 - All delegation addresses may submit votes at any point which updates their 
   preferred representation on the dCERT group. 
 - Preferred representation may be arbitrarily split between addresses (ex. 50%
   to John, 25% to Sally, 25% to Carol) 
 - Addresses which control the greatest amount of preferred-representation are
   eligible to join the dCERT group (up the _maximum number of dCERT members_)
   - In the split situation where the dCERT group is full but a vying candidate 
     has the same amount of vote as an existing dCERT member, the existing 
     member should maintain its position. 
 - In order for a new member to be added to the dCERT group they must 
   send a transaction accepting their admission at which point the validity of
   their admission is to be confirmed. 
   - A sequence number is assigned when a member is added to dCERT group. 
     If a member leaves the dCERT group and then enters back, a new sequence number
     is assigned.  

>  - If the dCERT group is already full and new member is admitted, the existing

Reviewer 2: Kind of a dup of the 2nd previous point. Maybe combine them?

Author: Good point. I merged these two together.
